### PR TITLE
Update run start and stop message schema

### LINKED ISSUE #106: Send file writer start and stop commands
After #34 is implemented it will be straightforward to send file writer start and stop messages for each run. This would be nice for testing as can go input file to output file and compare.

Wait until Mantid and file writer commands are being unified, this is in progress.

### PR DESCRIPTION
**Pull request by matthew-d-jones**
### Description of work

Updated to new run start and stop schema.
Updated dependencies to our current recommended versions.

### Issue

DM-1846
Closes #106

### CODE REVIEW

**Comment by mchorazak** [comment]
# [Codecov](https://codecov.io/gh/ess-dmsc/NeXus-Streamer/pull/117?src=pr&el=h1) Report
> Merging [#117](https://codecov.io/gh/ess-dmsc/NeXus-Streamer/pull/117?src=pr&el=desc) into [master](https://codecov.io/gh/ess-dmsc/NeXus-Streamer/commit/e0bad0125aa24b50f5274aa802ed3e2c9870ffcc&el=desc) will **increase** coverage by `0.17%`.
> The diff coverage is `n/a`.


```diff
@@            Coverage Diff             @@
##           master     #117      +/-   ##
==========================================
+ Coverage   90.01%   90.18%   +0.17%     
==========================================
  Files          37       38       +1     
  Lines         991      998       +7     
==========================================
+ Hits          892      900       +8     
+ Misses         99       98       -1     
```

| [Impacted Files](https://codecov.io/gh/ess-dmsc/NeXus-Streamer/pull/117?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [serialisation/src/UUID.cpp](https://codecov.io/gh/ess-dmsc/NeXus-Streamer/pull/117/diff?src=pr&el=tree#diff-c2VyaWFsaXNhdGlvbi9zcmMvVVVJRC5jcHA=) | `100.00% <0.00%> (ø)` | |

File: nexus_producer/src/NexusPublisher.cpp [hunk 1]
```diff
@@ -101,13 +98,16 @@ RunData
 NexusPublisher::createRunMessageData(const int runNumber,
                                      const std::string &jsonDescription) {
   auto runData = RunData();
-  runData.setNumberOfPeriods(m_fileReader->getNumberOfPeriods());
-  runData.setInstrumentName(m_fileReader->getInstrumentName());
-  runData.setRunID(std::to_string(runNumber));
+  runData.startTime = static_cast<uint64_t>(getTimeNowInNanoseconds());
```
Comment (mattclarke) [inline nexus_producer/src/NexusPublisher.cpp]: Everywhere getTimeNowInNanoseconds is used it gets cast to uint64_t, does it make sense to do the cast in the function?

File: serialisation/src/DetectorSpectrumMapData.cpp [hunk 2]
```diff
@@ -23,6 +24,17 @@ DetectorSpectrumMapData::DetectorSpectrumMapData(const std::string &filename) {
   readFile(filename);
 }
 
+DetectorSpectrumMapData::DetectorSpectrumMapData(
+    const SpectraDetectorMapping *detSpecMapFromMessage) {
+  auto detFBVector = detSpecMapFromMessage->detector_id();
```
Comment (mattclarke) [inline serialisation/src/DetectorSpectrumMapData.cpp]:
```suggestion
  const auto detFBVector = detSpecMapFromMessage->detector_id();
```

File: serialisation/src/DetectorSpectrumMapData.cpp [hunk 3]
```diff
@@ -23,6 +24,17 @@ DetectorSpectrumMapData::DetectorSpectrumMapData(const std::string &filename) {
...
+  auto specFBVector = detSpecMapFromMessage->spectrum();
```
Comment (mattclarke) [inline serialisation/src/DetectorSpectrumMapData.cpp]:
```suggestion
  const auto specFBVector = detSpecMapFromMessage->spectrum();
```

File: serialisation/src/DetectorSpectrumMapData.cpp [hunk 4]
```diff
@@ -23,6 +24,17 @@ DetectorSpectrumMapData::DetectorSpectrumMapData(const std::string &filename) {
...
+  setNumberOfEntries(detSpecMapFromMessage->n_spectra());
+  m_detectors.resize(static_cast<size_t>(m_numberOfEntries));
+  m_spectra.resize(static_cast<size_t>(m_numberOfEntries));
+  std::copy(detFBVector->begin(), detFBVector->end(), m_detectors.begin());
```
Comment (mattclarke) [inline serialisation/src/DetectorSpectrumMapData.cpp]:
```suggestion
  std::copy(detFBVector->cbegin(), detFBVector->cend(), m_detectors.begin());
```

File: serialisation/src/DetectorSpectrumMapData.cpp [hunk 5]
```diff
@@ -23,6 +24,17 @@ DetectorSpectrumMapData::DetectorSpectrumMapData(const std::string &filename) {
...
+  std::copy(specFBVector->begin(), specFBVector->end(), m_spectra.begin());
```
Comment (mattclarke) [inline serialisation/src/DetectorSpectrumMapData.cpp]:
```suggestion
  std::copy(specFBVector->cbegin(), specFBVector->cend(), m_spectra.begin());
```

File: serialisation/src/RunData.cpp [hunk 6]
```diff
@@ -34,69 +22,90 @@ uint64_t RunData::timeStringToUint64(const std::string &inputTime) {
...
 
-  auto instrumentName = builder.CreateString(m_instrumentName);
-  auto runID = builder.CreateString(m_runID);
-  auto nexusStructure = builder.CreateString(m_nexusStructure);
-  auto messageRunStart =
-      CreateRunStart(builder, m_startTime, runID, instrumentName,
-                     m_numberOfPeriods, nexusStructure);
-  auto messageRunInfo =
-      CreateRunInfo(builder, InfoTypes::RunStart, messageRunStart.Union());
+  auto instrumentName = builder.CreateString(runData.instrumentName);
```
Comment (mattclarke) [inline serialisation/src/RunData.cpp]: I'm guessing these all could be const too?

File: serialisation/src/RunData.cpp [hunk 7]
```diff
@@ -34,69 +22,90 @@ uint64_t RunData::timeStringToUint64(const std::string &inputTime) {
...
 
-Streamer::Message RunData::getRunStopBuffer() {
+Streamer::Message serialiseRunStopMessage(const RunData &runData) {
   flatbuffers::FlatBufferBuilder builder;
 
-  auto runID = builder.CreateString(m_runID);
-  auto messageRunStop = CreateRunStop(builder, m_stopTime, runID);
-  auto messageRunInfo =
-      CreateRunInfo(builder, InfoTypes::RunStop, messageRunStop.Union());
+  auto runID = builder.CreateString(runData.runID);
```
Comment (mattclarke) [inline serialisation/src/RunData.cpp]: const?

**Pull request merged**